contributors/design-proposals/resource-class.md

```diff
@@ -0,0 +1,363 @@
+   such resource classes, update the availability state in the local cache.
+3. ResourceClass api object's status, `ResourceClassStatus` will be patched
+   as per the updated availability state in the cache.
+
+### Node Deletion
+If node has devices which are selectable by existing resource classes:
+1. Adjust resource class state in local cache.
+2. Update resource class status by patching api object.
+
+### Pod Creation
+1. Get the requested resource class name and quantity from pod spec.
+2. Select nodes by applying predicates according to requested quantity and Resource
+   class's state present in the cache.
+3. On the selected node, select a Device from the stored devices info in cache
+   after matching key,value from requested resource class.
+4. After device selection, update(decrease) 'Requested' for all the resource
+   classes which could select this device in the cache.
+5. Patch the resource class objects with new 'Requested' in the `ResourceClassStatus`.
+6. Add the pod reference in local DeviceToPod mapping structure in the cache.
+7. Patch the pod object with selected device annotation with prefix 'scheduler.alpha.kubernetes.io/resClass'
+
+NOTE: This proposal propose only 'first fit' as device selection strategy.
+      In the future, this can be extended to multiple algorithms available for
+      the user to choose from, in a configurable manner.
+
+### Pod Delete
+1. Iterate over the all the devices on the at which pod was scheduled to and
+   find out the devices being used by pod.
+2. For each device consumed by pod, update availability state of Resource classes
+   which can select this device in the cache.
+3. Patch `ResourceClassStatus` with new availability state.
+
+## Kubelet Changes
+Update logic at container runtime manager to look for device annotations,
+prefixed by 'scheduler.alpha.kubernetes.io/resClass' and call matching device
+plugins.
+
+## Opaque Integer Resources
+This API will supersede the [Opaque Integer Resources](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#opaque-integer-resources-alpha-feature)
+(OIR). External agents can continue to attach additional 'opaque' resources to
+nodes, but the special naming scheme that is part of the current OIR approach
+will no longer be necessary. Any existing resource discovery tool which updates
+node objects with OIR, will adapt to update node status with devices instead.
+
+
+## Future Scope
+* RBAC: It can further be explored that how to tie resource classes with RBAC
+  like any other existing API resource objects.
+* Nested Resource Classes: In future device plugins and resource classes can be 
+  extended to support the nested resource class functionality where one resource
+  class could be comprised of a group of sub-resource classes. For example 'numa-node'
+  resource class comprised of sub-resource classes, 'single-core'.
+* Multiple device selection algorithms, each with a different selection strategy,
```

Comment:
I think this should be thoroughly discussed and not just a "side note" on the bottom of the design doc.
The decision we take regarding the way we address this problem will have lasting impact on
GPU providers (e.g: GKE).
But then I'm not against selling more GPUs because we have poor scheduling algorithms :D